crates/rpc/rpc/src/eth/api/state.rs

```diff
@@ -31,9 +33,32 @@ where
         address: Address,
         block_id: Option<BlockId>,
     ) -> EthResult<U256> {
-        let state = self.state_at_block_id_or_latest(block_id)?;
-        let nonce = U256::from(state.account_nonce(address)?.unwrap_or_default());
-        Ok(nonce)
+        match block_id {
+            Some(BlockId::Number(BlockNumberOrTag::Pending)) => {
+                let address_txs = self.pool().get_all_transactions_by_sender(address);
+                // get max nonce
+                let highest_nonce = address_txs.iter().reduce(|accum, item| {
+                    let nonce = item.transaction.nonce();
+                    if nonce.gt(&accum.transaction.nonce()) {
+                        item
+                    } else {
+                        accum
+                    }
+                });
+
+                let on_chain_nonce = if let Some(highest_nonce_tx) = highest_nonce {
+                    U256::from(highest_nonce_tx.transaction.nonce())
+                } else {
+                    U256::from(0)
+                };
+
+                EthResult::Ok(on_chain_nonce)
```

Comment:
this needs to be an if let and should not return if `address_txs` is empty, instead it should go into the other arm.